api-metastore/src/main/java/org/zalando/nakadi/service/EventTypeService.java

```diff
@@ -398,34 +398,42 @@ public void delete(final String eventTypeName) throws EventTypeDeletionException
-                SubscriptionTokenLister.ListResult listResult = subscriptionTokenLister.listSubscriptions(
-                        ImmutableSet.of(eventType), Optional.empty(), Optional.empty(), null, 100);
-                while (null != listResult) {
-                    listResult.getItems().forEach(s -> {
-                        try {
-                            subscriptionRepository.deleteSubscription(s.getId());
-                        } catch (final NoSuchSubscriptionException e) {
-                            // should not happen as we are inside transaction
-                            throw new InconsistentStateException("Subscription to be deleted is not found", e);
+            return eventTypeRepository.lockingTable(EventTypeRepository.TableLockMode.ROW_EXCLUSIVE,
+                    transactionTemplate,
+                    action -> {
+                        SubscriptionTokenLister.ListResult listResult = listSubscriptions(eventType, null, 100);
+                        while (null != listResult) {
+                            listResult.getItems().forEach(s -> {
+                                try {
+                                    subscriptionRepository.deleteSubscription(s.getId());
+                                } catch (final NoSuchSubscriptionException e) {
+                                    // should not happen as we are inside transaction
+                                    throw new InconsistentStateException("Subscription to be deleted is not found", e);
+                                }
+                            });
+                            listResult = null == listResult.getNext() ? null :
+                                    listSubscriptions(eventType, listResult.getNext(), 100);
                         }
-                    });
-                    listResult = null == listResult.getNext() ? null : subscriptionTokenLister.listSubscriptions(
-                            ImmutableSet.of(eventType), Optional.empty(), Optional.empty(), listResult.getNext(), 100);
-                }
-                return deleteEventType(eventType);
-            });
+                        return deleteEventType(eventType);
+                    }
+            );
+
         } catch (final TransactionException e) {
             throw new InconsistentStateException("Failed to delete event-type because of race condition in DB", e);
         }
     }
 
+    private SubscriptionTokenLister.ListResult listSubscriptions(final String eventType,
+                                                                 final SubscriptionTokenLister.Token token,
+                                                                 final int limit) {
+        return subscriptionTokenLister.listSubscriptions(
+                ImmutableSet.of(eventType), Optional.empty(), Optional.empty(), token, limit);
+    }
 
     // TODO: This method should be fixed by creating proper db query.
```

Comment:
I've just noticed this TODO.  Should we check if we can fix it in the course of this change?

Reply:
Ill make a separate PR for this.

Reply:
Why?  I think this is well in scope of the change with ensuring DB consistency.

Reply:
The ticket is about "Db consistency" correct but the problem statement is about orphan subscription. The todo deals with nakadi settings as far as i see. Also why nakadi settings should be fetched from DB is another question I have.

Reply:
I think I understood the problem (not related to fetch nakadi settings from db) now but would still prefer to make a separate PR

Reply:
OK, I've had a closer look as well and I agree that this can be handled separately.  No dangling subscriptions can result from lack of transaction here (but some subscriptions may be deleted when they shouldn't be).